### ISSUE TITLE
A problem during training

### ISSUE THREAD

**Issue by Xinzhuo1** [stefanos50/CARLA2Real#1]
Hallo
Thanks for the great work. i am trying to start the training procedure，but something went wrong.This is the problem of the code or the problem caused by insufficient video memory and memory？

![Image](https://github.com/user-attachments/assets/d18635eb-d862-4d54-8f39-866ceebe4367)

Thanks for your help!

**Comment by stefanos50** [comment]
> Hallo Thanks for the great work. i am trying to start the training procedure，but something went wrong.This is the problem of the code or the problem caused by insufficient video memory and memory？
> 
> ![Image](https://github.com/user-attachments/assets/d18635eb-d862-4d54-8f39-866ceebe4367)
> 
> Thanks for your help!

Hi, is PyTorch build correctly with GPU support and the required CUDA library? I would recommend using a conda environment. Can you also specify the GPU model you are trying to run the training? Are you using the dataset from kaggle to perform the training? There are also pretrained models available that you can use. To train you should also read all the documentation and the required steps from the EPE repository. The kaggle dataset is already preprocessed and you will only need the MSEG results from both the synthetic data and the real-world images.

**Comment by Xinzhuo1** [comment]
> > Hallo Thanks for the great work. i am trying to start the training procedure，but something went wrong.This is the problem of the code or the problem caused by insufficient video memory and memory？
> > ![Image](https://github.com/user-attachments/assets/d18635eb-d862-4d54-8f39-866ceebe4367)
> > Thanks for your help!
> 
> Hi, is PyTorch build correctly with GPU support and the required CUDA library? I would recommend using a conda environment. Can you also specify the GPU model you are trying to run the training? Are you using the dataset from kaggle to perform the training? There are also pretrained models available that you can use. To train you should also read all the documentation and the required steps from the EPE repository. The kaggle dataset is already preprocessed and you will only need the MSEG results from both the synthetic data and the real-world images.

Thanks for your help!I found the problem but don't know how to fix it
By the way, I was successful in the test procedure but had a problem with the training procedure, so I don't think it was the GPU support or CUDA library that caused the problem
I used the cpu for training to display the full error message and the gpu did not display the specific error message. The error message（cpu） is as follows

![Image](https://github.com/user-attachments/assets/7ca8d946-69b3-4e33-bd9a-fa6f16bb7ba3)
The wrong location is here.The problem is caused by embedding，Look at line 170.

![Image](https://github.com/user-attachments/assets/29554f49-390c-4aed-9090-0b16d19248f8)

![Image](https://github.com/user-attachments/assets/87ca70e2-a757-4312-9c63-0f552cb0380c)

I used universal_demo.py of Mseg to get the grayscale image, but some pixel values are greater than 194, so how should I solve it? Should I modify the grayscale so that all pixel values are less than 194 or should I modify num_embeddings (more than 194 classes that is used in the disciminator) or use some other appropriate segmentation network？
Do you also use segmentations generated with MSeg as robust label maps and have no problems？

**Comment by stefanos50** [comment]
> > > Hallo Thanks for the great work. i am trying to start the training procedure，but something went wrong.This is the problem of the code or the problem caused by insufficient video memory and memory？
> > > ![Image](https://github.com/user-attachments/assets/d18635eb-d862-4d54-8f39-866ceebe4367)
> > > Thanks for your help!
> > 
> > 
> > Hi, is PyTorch build correctly with GPU support and the required CUDA library? I would recommend using a conda environment. Can you also specify the GPU model you are trying to run the training? Are you using the dataset from kaggle to perform the training? There are also pretrained models available that you can use. To train you should also read all the documentation and the required steps from the EPE repository. The kaggle dataset is already preprocessed and you will only need the MSEG results from both the synthetic data and the real-world images.
> 
> Thanks for your help!I found the problem but don't know how to fix it By the way, I was successful in the test procedure but had a problem with the training procedure, so I don't think it was the GPU support or CUDA library that caused the problem I used the cpu for training to display the full error message and the gpu did not display the specific error message. The error message（cpu） is as follows
> 
> ![Image](https://github.com/user-attachments/assets/7ca8d946-69b3-4e33-bd9a-fa6f16bb7ba3) The wrong location is here.The problem is caused by embedding，Look at line 170.
> 
> ![Image](https://github.com/user-attachments/assets/29554f49-390c-4aed-9090-0b16d19248f8)
> 
> ![Image](https://github.com/user-attachments/assets/87ca70e2-a757-4312-9c63-0f552cb0380c)
> 
> I used universal_demo.py of Mseg to get the grayscale image, but some pixel values are greater than 194, so how should I solve it? Should I modify the grayscale so that all pixel values are less than 194 or should I modify num_embeddings (more than 194 classes that is used in the disciminator) or use some other appropriate segmentation network？ Do you also use segmentations generated with MSeg as robust label maps and have no problems？



The error is related to the original EPE code (something i personally didnt modified) so it should be connected with the dataset.  I remember an issue in the EPE repository with MSEG stating that indeed the num of classes for MSEG are hardcoded (https://github.com/isl-org/PhotorealismEnhancement/issues/64).

I have the robust (MSEG) images for the synthetic dataset you can download them here (https://drive.google.com/file/d/1q9Ja6b1mOn_LioDGU-NCYIOKKgDa8wl2/view?usp=sharing) and check if there is any difference with the ones you generated.

The data loading happens in /code/epe/dataset/pdf.py, i suggest to check if the variables img/gbuffers/gt_labels/robust_labels are all loaded correctly.

Additionally, these are the commands i used to generate the MSEG label maps:

-----MSEG Robust Semantic Segmentation----

python3 -u universal_demo.py --config=/mnt/g/mseg-semantic/mseg_semantic/config/test/default_config_360_ms.yaml --model_name=mseg-3m --model_path=/mnt/g/mseg-semantic/mseg_semantic/trained_models/mseg-3m-1080p.pth

wsl #open subsystem
cd /mnt/g/mseg-semantic/mseg_semantic/tool
config=/mnt/g/mseg-semantic/mseg_semantic/config/test/default_config_360_ms.yaml
model_name=mseg-3m or model_name=cityscapes-19-1m
model_path=/mnt/g/mseg-semantic/mseg_semantic/trained_models/mseg-3m-1080p.pth or model_path=/mnt/g/mseg-semantic/mseg_semantic/trained_models/cityscapes-19-1m.pth
input_file=/mnt/a/CityScapesDataset/CityscapesImages or input_file=/mnt/a/MapillaryVistasDataset/MapillaryVistasImages
save_folder=/mnt/a/CityScapesDataset or save_folder=/mnt/a/MapillaryVistasDataset

Which real-world dataset you are trying to use?

**Comment by stefanos50** [comment]
If i remember correct i used the mseg-3m (mseg-3m-1080p.pth) pretrained model. Its likely that the different models that are provided in the MSEG repo will output different number of classes. Since the classes are hardcoded for a specific model may this is the issue.

**Comment by Xinzhuo1** [comment]
> If i remember correct i used the mseg-3m (mseg-3m-1080p.pth) pretrained model. Its likely that the different models that are provided in the MSEG repo will output different number of classes. Since the classes are hardcoded for a specific model may this is the issue.如果我没记错的话，我使用了 mseg-3m （mseg-3m-1080p.pth） 预训练模型。MSEG 存储库中提供的不同模型可能会输出不同数量的类。由于类是针对特定模型进行硬编码的，因此这可能是问题所在。

I tried the mseg-3m-1080p.pth and cityscapes-19-1m.pth but some pixel values are still greater than 194.
I'm going to use your data for training because I found the robust (MSEG) images for the synthetic dataset was appropriate，  or i will modify the my own grayscale image.
Thank you for your help，best wishes for you！

**Comment by stefanos50** [comment]
> > If i remember correct i used the mseg-3m (mseg-3m-1080p.pth) pretrained model. Its likely that the different models that are provided in the MSEG repo will output different number of classes. Since the classes are hardcoded for a specific model may this is the issue.如果我没记错的话，我使用了 mseg-3m （mseg-3m-1080p.pth） 预训练模型。MSEG 存储库中提供的不同模型可能会输出不同数量的类。由于类是针对特定模型进行硬编码的，因此这可能是问题所在。
> 
> I tried the mseg-3m-1080p.pth and cityscapes-19-1m.pth but some pixel values are still greater than 194. I'm going to use your data for training because I found the robust (MSEG) images for the synthetic dataset was appropriate， or i will modify the my own grayscale image. Thank you for your help，best wishes for you！

I reinstalled MSEG and tried these models:

1) https://drive.google.com/file/d/1iobea9IW2cWPF6DtM04OvKtDRVNggSb5/view
2) https://drive.google.com/file/d/1iobea9IW2cWPF6DtM04OvKtDRVNggSb5/view

All the predicted labels were under 194. You will also need to extract the MSEG results for your real-world dataset.
